### PR TITLE
Move horizontal scroll when it required by selection

### PR DESCRIPTION
**Pull request by Kvaz1r**
For https://github.com/nelsam/vidar/issues/136

If this approach good enough I also can add analogic addes for vertical scroll.

### CODE REVIEW

File: mixins/textbox.go [hunk 1]
```diff
@@ -585,6 +585,25 @@ func (t *TextBox) MouseMove(ev gxui.MouseEvent) {
...
+			lineOffset := t.lineWidthOffset()
+			padding := t.Padding()
+			horizStart := t.horizOffset
+			horizEnd := t.horizOffset + size.W - padding.W() - lineOffset
+			line, _ := t.outer.CreateLine(t.theme, lineIndex)
+			pos := line.PositionAt(p)
+
+			from, to := t.horizScroll.ScrollPosition()
+			L := to - from
+			delta := 10
```
Comment (nelsam) [inline mixins/textbox.go]: I'd like this to be a constant, e.g.
```go
const SelectionScrollBuffer = 10`
```

**Comment by nelsam** [comment]
On my computer, this scrolls *very* fast.  I'd like to see it edited with something like the following:

```go
const MaxSelectionScrollRate = 100 * time.Millisecond
now := time.Now()
if now.Before(t.lastSelectionScroll.Add(MaxSelectionScrollRate)) {
    return
}
t.lastSelectionScroll = now
```

I also noticed that this only scrolls when you move your mouse ... is that how we want it to work?  We could alternatively have the scrolling start as soon as the mouse moves in to that buffer zone, then stop as soon as it moves out, e.g.

```go
if t.stopScrolling != nil {
    return
}
t.stopScrolling = make(chan struct{})
go func() {
    ticker := time.NewTicker(MaxSelectionScrollRate)
    defer func() {
        ticker.Stop()
        t.stopScrolling = nil
    }()
    for range ticker.C {
        select {
        case <-t.stopScrolling:
            return
        default:
            t.updateSelectionScroll()
        }
    }
}()
```

^ in that code, we would need to run `close(t.stopScrolling)` when the mouse moves away from the edge of the editor.

**Comment by Kvaz1r** [comment]
> On my computer, this scrolls very fast. I'd like to see it edited with something like the following:

Agree, but not sure that it would be enough. Maybe excess of selection - also consequence of the fact that mouse don't move. I am going to check it.  

> I also noticed that this only scrolls when you move your mouse ... is that how we want it to work?

I think so. I don't see many difference between these approaches for lines with long ~ 2 x width of screen and this is a main case when scrolling needed. For longer lines absence of explicit action will give selection to end (start) of line, isn't?

**Comment by nelsam** [comment]
I played with your changes for a while and came up with something based on what you were doing, but with a few changes to how the selection scrolling works.  It's not completely done, but it shows an adjustable scroll rate (it scrolls faster as your mouse gets closer to the edge) and constant scrolling while your mouse stays still.

**Comment by Kvaz1r** [comment]
Great, I quickly tested your solution and it works very good.

**Pull request merged**